nvflare/private/fed/utils/fed_utils.py

```diff
@@ -236,3 +236,38 @@ def create_stats_pool_files_for_job(workspace: Workspace, job_id: str, prefix=No
         err = f"Failed to create stats pool summary file {summary_file}: {secure_format_exception(e)}"
     StatsPoolManager.close()
     return err
+
+
+def split_gpus(gpus) -> ([str], bool):
```

Comment:
This should work, but parsing seems to be tedious. Consider using something like this:


def find_char(s, ch):
    return [i for i, c in enumerate(s) if c == ch]


def split_gpus(gpus) -> [str]:
    s = gpus.replace(" ", "")
    lefts = find_char(s, '[')
    rights = find_char(s, ']')
    if len(lefts) != len(rights):
        raise ValueError('brackets not paired')

    offset = 0
    for i in range(len(lefts)):
        if lefts[i] > rights[i]:
            raise ValueError('brackets not properly paired')
        p = s[lefts[i]+1-offset:rights[i]-offset]
        p = p.replace(',', ':')
        s = s.replace(s[lefts[i]-offset:rights[i]+1-offset], p)
        offset += 2

    result = s.split(',')
    for i in range(len(result)):
        result[i] = result[i].replace(':', ',')
    print(f"{gpus} ==> {result}")
    return result

Reply:
Somehow the indentation got messed up, but you should be able to see it. The idea is to first find all pairing brackets. Then for the content in each pair, replace "," with ":"; and then put the replaced content back to the original list. This would reduce the total str size by 2 for each pair. After this process, the groups are only separated by ",".  Then you replace ":" with "," for each group.

This works for all your test cases, and it also works for syntax like "1:2:3, [4,5], [3:2], 1, 5".
It will also catch improperly paired brackets like:
"[1,2,3"
"]1, 2, 3["

Reply:
I am not sure if "replace(',', ':')" is a good idea. This is a split gpus string function. In theory the GPU ID / name could be also contain ":" character as well, depending on different system. So maybe change the content of provided GPU name may break it.

Reply:
In that case, we could use a different char than ":", say "#" or "@". Idea is to make it easier for split with comma later.

Reply:
I also updated the code to protect against bad start or end of a group and nested brackets.